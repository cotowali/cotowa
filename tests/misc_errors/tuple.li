fn ...(int, int) |> f() |> ...(int, int) {
  return 0
  return ('x', 'y')
}

0 |> f()

fn () |> zero() |> int {
  var v: int
  read(&v)
  return 0
}

0 |> zero()
(1, 2) |> zero()

(0, 0) == (1)
(0, 0) == ('a', 'b')
(0, 0) * (0, 0)
(0, 0) / (0, 0)
(0, 1) + 2

var v: (...int)
var nested: ((int, int))

fn non_tuple_decompose() {
  var n = 10
  var v: (int, int)
  v = (...n, ...n)
  v = (...0, ...0)
}

fn tuple_index() {
  var n = 10

  _ = (0, "10")[-1]
  _ = (0, "10")[2]
  _ = (0, "10")[n]
  var _: int = (0, "10")[1]

  type IntStr = (int, string)
  var int_str: IntStr = (0, "10")
  _ = int_str[-1]
  _ = int_str[2]
  _ = int_str[n]
  var _: int = int_str[1]
}
