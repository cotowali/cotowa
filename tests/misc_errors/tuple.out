error: tuple.li 2,10: mismatched types: `...(int, int)` (expected) and `int` (found)
    2 |   return 0
      |          ^
error: tuple.li 3,10: mismatched types: `...(int, int)` (expected) and `(string, string)` (found)
    3 |   return ('x', 'y')
      |          ^^^^^^^^^^
error: tuple.li 6,1: mismatched types: `int` (left) and `(int, int)` (pipe in of right)
    6 | 0 |> f()
      | ^^^^^^^^
error: tuple.li 10,3: expected 0 arguments, but got 1
   10 |   read(&v)
      |   ^^^^^^^^
error: tuple.li 14,1: mismatched types: `int` (left) and `()` (pipe in of right)
   14 | 0 |> zero()
      | ^^^^^^^^^^^
error: tuple.li 15,1: mismatched types: `(int, int)` (left) and `()` (pipe in of right)
   15 | (1, 2) |> zero()
      | ^^^^^^^^^^^^^^^^
error: tuple.li 17,1: mismatched types: `(int, int)` (left) and `int` (right)
   17 | (0, 0) == (1)
      | ^^^^^^^^^^^^^
error: tuple.li 18,1: mismatched types: `(int, int)` (left) and `(string, string)` (right)
   18 | (0, 0) == ('a', 'b')
      | ^^^^^^^^^^^^^^^^^^^^
error: tuple.li 19,1: invalid operation: `(int, int)` * `(int, int)`
   19 | (0, 0) * (0, 0)
      | ^^^^^^^^^^^^^^^
error: tuple.li 20,1: invalid operation: `(int, int)` / `(int, int)`
   20 | (0, 0) / (0, 0)
      | ^^^^^^^^^^^^^^^
error: tuple.li 21,1: mismatched types: `(int, int)` (left) and `int` (right)
   21 | (0, 1) + 2
      | ^^^^^^^^^^
error: tuple.li 23,9: cannot use sequence type as tuple element
   23 | var v: (...int)
      |         ^^^^^^
error: tuple.li 24,14: cannot use tuple as tuple element
   24 | var nested: ((int, int))
      |              ^^^^^^^^^^
error: tuple.li 29,8: cannot decompose non-tuple type `int`
   29 |   v = (...n, ...n)
      |        ^^^^
error: tuple.li 29,14: cannot decompose non-tuple type `int`
   29 |   v = (...n, ...n)
      |              ^^^^
error: tuple.li 30,8: cannot decompose non-tuple type `int`
   30 |   v = (...0, ...0)
      |        ^^^^
error: tuple.li 30,14: cannot decompose non-tuple type `int`
   30 |   v = (...0, ...0)
      |              ^^^^
error: tuple.li 36,17: index -1 out of bounds for `(int, string)` (2 elements tuple)
   36 |   _ = (0, "10")[-1]
      |                 ^^
error: tuple.li 37,17: index 2 out of bounds for `(int, string)` (2 elements tuple)
   37 |   _ = (0, "10")[2]
      |                 ^
error: tuple.li 38,17: index of tuple must be int literal.
   38 |   _ = (0, "10")[n]
      |                 ^
error: tuple.li 43,15: index -1 out of bounds for `IntStr` (2 elements tuple)
   43 |   _ = int_str[-1]
      |               ^^
error: tuple.li 44,15: index 2 out of bounds for `IntStr` (2 elements tuple)
   44 |   _ = int_str[2]
      |               ^
error: tuple.li 45,15: index of tuple must be int literal.
   45 |   _ = int_str[n]
      |               ^
